servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/P2CSelector.java

```diff
@@ -0,0 +1,136 @@
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.context.api.ContextMap;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Single.succeeded;
+
+/**
+ * This {@link LoadBalancer} selection algorithm is based on work by Michael David Mitzenmacher in The Power of Two
+ * Choices in Randomized Load Balancing.
+ *
+ * @see <a href="https://www.eecs.harvard.edu/~michaelm/postscripts/tpds2001.pdf">Mitzenmacher (2001) The Power of Two
+ * Choices in Randomized Load Balancing</a>
+ */
+final class P2CSelector<ResolvedAddress, C extends LoadBalancedConnection>
+        extends BaseHostSelector<ResolvedAddress, C> {
+
+    @Nullable
+    private final Random random;
+    private final int maxEffort;
+
+    P2CSelector(final String targetResource, final int maxEffort, @Nullable final Random random) {
+        super(targetResource);
+        this.maxEffort = maxEffort;
+        this.random = random;
+    }
+
+    @Override
+    public Single<C> selectConnection(
+            @Nonnull List<Host<ResolvedAddress, C>> hosts,
+            @Nonnull Predicate<C> selector,
+            @Nullable ContextMap context,
+            boolean forceNewConnectionAndReserve) {
+        final int size = hosts.size();
+        switch (size) {
+            case 0:
+                // We shouldn't get called if the load balancer doesn't have any hosts.
+                throw new AssertionError("Selector for " + getTargetResource() +
+                        " received an empty host set");
+            case 1:
+                Host<ResolvedAddress, C> host = hosts.get(0);
+                if (!forceNewConnectionAndReserve) {
+                    C connection = host.pickConnection(selector, context);
+                    if (connection != null) {
+                        return succeeded(connection);
+                    }
+                }
+                // Either we require a new connection or there wasn't one already established so
+                // try to make a new one if the host is healthy. If it's not healthy, we fail
+                // and let the higher level retries decide what to do.
+                if (!host.isActiveAndHealthy()) {
+                    return noActiveHostsException(hosts);
+                }
+                return host.newConnection(selector, forceNewConnectionAndReserve, context);
+            default:
+                return p2c(size, hosts, getRandom(), selector, forceNewConnectionAndReserve, context);
+        }
+    }
+
+    @Nullable
+    private Single<C> p2c(int size, List<Host<ResolvedAddress, C>> hosts, Random random, Predicate<C> selector,
+                          boolean forceNewConnectionAndReserve, @Nullable ContextMap contextMap) {
+        for (int j = maxEffort; j > 0; j--) {
```

Comment:
Will it make sense to limit efforts based on number of hosts? For example, if there are only 2 hosts with `maxEffort > 1`, should we try again or fail fast?